File: src/chared/asset.ts
import {Box, Dim, boxIn, boxGrow} from 'game/geo'
import {Grid, gridTiles, gridEach} from 'game/grid'

export const kind:{[kind:string]:string} = {
	char:'Character',
	tile:'Map Tiles',
	item:'Items',
	icon:'Icons',
}
export const kinds = Object.keys(kind).map(k => ({kind:k, name:kind[k]}))

export type Pixel = number

export interface AssetInfo {
	name:string
	kind:string
}

export interface Sequence {
	name:string
	ids:number[]
}

export interface Asset extends AssetInfo, Dim {
	seq:Sequence[]
	pal:string
	pics:{[id:PicID]:Pic}
}


export interface Sel extends Grid<Pixel> { }

export type PicID = number
export interface Pic extends Grid<Pixel> { id:PicID }

export function picInit(p:Pic):Pic {
	return Object.assign(p, gridTiles<Pixel>(p, p.raw))
}

export interface EditPic extends Sel {
	pic:PicID
	copy?:boolean
}

export function growPic(p:Grid<Pixel>, o:Box) {
	if (o.w*o.h<=0||boxIn(o, p)) return
	const b = boxGrow(p, o)
	let tmp = gridTiles<Pixel>(b)
	gridEach(p, (p, t) => tmp.set(p, t))
	Object.assign(p, tmp)
}

File: game/pix/asset.go
package pix

import (
	"github.com/mb0/babasite/game/geo"
	"github.com/mb0/babasite/game/ids"
)

type AssetInfo struct {
	Name string `json:"name"`
	Kind string `json:"kind"`
	geo.Dim
	Seq []*Seq `json:"seq"`
	Pal string `json:"pal"`
}

func (a *AssetInfo) GetSeq(name string) *Seq {
	for _, s := range a.Seq {
		if s.Name == name {
			return s
		}
	}
	return nil
}

func (a *AssetInfo) AddSeq(name string) *Seq {
	if s := a.GetSeq(name); s != nil {
		return s
	}
	s := &Seq{Name: name}
	a.Seq = append(a.Seq, s)
	return s
}

type Asset struct {
	AssetInfo
	Pics map[ids.Pic]*Pic `json:"pics"`
	Last ids.Pic          `json:"-"`
}

func NewAsset(info AssetInfo) *Asset {
	return &Asset{AssetInfo: info, Pics: make(map[ids.Pic]*Pic)}
}

func (a *Asset) GetPics(is ...ids.Pic) []*Pic {
	pics := make(map[ids.Pic]bool, len(is))
	res := make([]*Pic, 0, len(is))
	for _, id := range is {
		_, ok := pics[id]
		if !ok {
			pics[id] = true
			res = append(res, a.Pics[id])
		}
	}
	return res
}

func (a *Asset) NewPic() *Pic {
	a.Last += 1
	p := &Pic{ID: a.Last}
	a.Pics[p.ID] = p
	return p
}
